lib/IRGen/IRGenSIL.cpp

```diff
@@ -2059,33 +2102,126 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+            elementTypeString == "Bool" || elementTypeString == "String") {
+          assert(0 && "TODO: implement type list typed attr");
+        }
+        if (elementTypeString == "TensorShape" ||
+            elementTypeString == "Optional<TensorShape>") {
+          assert(0 && "TODO: implement shape list typed attr");
+        }
+        if (elementType->is<AnyMetatypeType>()) {
+          assert(0 && "TODO: implement type list typed attr");
+        }
+        assert(0 && "unknown array attribute");
+      }
+      case SymbolicValue::Integer:
+      case SymbolicValue::Float: {
+        auto attrNameAddr = createStringValAddr(IGM, attrName.c_str());
+        llvm::Constant *setAttrFn = nullptr;
+        llvm::Value *attrVal = nullptr;
+        if (attr.value.getKind() == SymbolicValue::Float) {
+          // TODO: confirm unit test coverage for float-typed attr.
+          setAttrFn = IGM.getTFE_OpSetAttrFloatFn();
+          auto value = attr.value.getFloatValue();
+          LLVM_DEBUG(llvm::dbgs() << " Setting float-typed attr " << attrName
+                                  << " with value " << attr.value << ".\n");
+          // TensorFlow only supports 32-bit float attributes.  If we got a 16
+          // or 64 bit one, convert it to float.
+          bool losesInfo = false;
+          value.convert(APFloat::IEEEsingle(), APFloat::rmNearestTiesToEven,
+                        &losesInfo);
+          attrVal = llvm::ConstantFP::get(IGM.FloatTy, value.convertToFloat());
+        } else {
+          auto value = attr.value.getIntegerValue();
+          LLVM_DEBUG(llvm::dbgs() << " Setting int-typed attr " << attrName
+                                  << " with value " << attr.value << ".\n");
+          if (value.getBitWidth() == 1) {
+            setAttrFn = IGM.getTFE_OpSetAttrBoolFn();
+            attrVal =
+                llvm::ConstantInt::get(IGM.Int1Ty, value.isAllOnesValue());
+          } else {
+            setAttrFn = IGM.getTFE_OpSetAttrIntFn();
+            attrVal =
+                llvm::ConstantInt::get(IGM.Int64Ty, value.getLimitedValue());
+          }
+        }
+        Builder.CreateCall(setAttrFn, {op, attrNameAddr, attrVal});
+        break;
+      }
+      case SymbolicValue::String: {
+        auto attrValue = attr.value.getStringValue();
+        if (attrName != "__device") {
+          auto attrNameAddr = createStringValAddr(IGM, attrName.c_str());
+          auto attrValAddr = createStringValAddr(IGM, attrValue);
+          auto attrLength =
+              llvm::ConstantInt::get(IGM.SizeTy, attrValue.size());
+          auto *setAttrFn = IGM.getTFE_OpSetAttrStringFn();
+          Builder.CreateCall(setAttrFn,
+                             {op, attrNameAddr, attrValAddr, attrLength});
+        } else {
           // In this case, just let eager pick a default device.
-          if (deviceStr == "ALL_DEVICES")
+          if (attrValue == ALL_DEVICES)
```

Comment:
The variable names for these device name constants ("ALL_DEVICES", "CPU", "GPU", and "TPU") are going to be confusing here because it's some TF-specific code in general IRGen. Consider adding a "TF_" prefix to these variable names.

Reply:
That sounds reasonable. I left a TODO, and will tackle it in another PR, so that this PR does not get largeer (and harder to review).